lib/shopify_app/session/session_storage.rb

```diff
@@ -3,9 +3,12 @@ module SessionStorage
     extend ActiveSupport::Concern
 
     included do
-      validates :shopify_domain, presence: true, uniqueness: { case_sensitive: false }
       validates :shopify_token, presence: true
       validates :api_version, presence: true
+      validates :shopify_domain, presence: true,
+        if: Proc.new {|_| ShopifyApp.configuration.per_user_tokens }
+      validates :shopify_domain, presence: true, uniqueness: { case_sensitive: false },
+        if: Proc.new {|_| !ShopifyApp.configuration.per_user_tokens }
```

Comment:
One is applied when we are using shop tokens, and the other when using user tokens. They have different validation requirements.

Reply:
Why would domain uniqueness matter only when using shop tokens?

Reply:
Because the opposite of per-user is per-shop. There would only be one session for each shop. The domain uniqueness is to ensure that there is only ever one per shop